src/txmempool.cpp

```diff
@@ -889,6 +982,140 @@ void CTxMemPool::removeForBlock(const std::vector<CTransactionRef> &vtx,
     bool fCurrentEstimate,
     std::vector<CTxChange> *txChanges)
+{
+    WRITELOCK(cs_txmempool);
+
+    // Process the block for transasction removal and ancestor state updates.
+    // As a first step remove all the txns that are in the block from the mempool by
+    // first updating the children for removal of the parent, and also saving the entries
+    // for policy estimate updates.  Then in the final loop simply remove all the txns.
+    //
+    // While we're updating the child transactions prior to removal we can also gather the
+    // mapTxnChaiTips.
+    mapEntryHistory mapTxnChainTips;
+    std::vector<CTxMemPoolEntry> entries;
+    setEntries setAncestorsFromBlock;
+    {
+        setEntries setTxnsInBlock;
+        for (const auto &tx : vtx)
+        {
+            uint256 hash = tx->GetHash();
+            txiter it = mapTx.find(hash);
+            if (it == mapTx.end())
+            {
+                continue;
+            }
+
+            // Get all ancestors from related to txns in the block. Stop looking if we've already looked up
+            // this set of ancestors before; we don't want to be traversing the same part of the ancestor
+            // tree more than once.
+            setEntries setAncestors;
+            uint64_t nNoLimit = std::numeric_limits<uint64_t>::max();
+            std::string dummy;
+            _CalculateMemPoolAncestors(
+                *it, setAncestors, nNoLimit, nNoLimit, nNoLimit, nNoLimit, dummy, &setAncestorsFromBlock, false);
+            setAncestorsFromBlock.insert(setAncestors.begin(), setAncestors.end());
+
+            setTxnsInBlock.insert(it);
+            entries.push_back(*it); // TODO: seems inefficient to do a full copy here?
+
+            const setEntries &setMemPoolChildren = GetMemPoolChildren(it);
+            for (txiter updateIt : setMemPoolChildren)
+            {
+                // Get the current ancestor state values and save them for later comparison
+                ancestor_state ancestorState(updateIt->GetSizeWithAncestors(), updateIt->GetModFeesWithAncestors(),
+                    updateIt->GetCountWithAncestors(), updateIt->GetSigOpCountWithAncestors());
+                mapTxnChainTips.emplace(updateIt, ancestorState);
+
+                // Remove each block entry from the parent map in mapLinks
+                _UpdateParent(updateIt, it, false);
+            }
+        }
+        for (txiter it : setTxnsInBlock)
+        {
+            setAncestorsFromBlock.erase(it);
+            mapTxnChainTips.erase(it);
+            removeUnchecked(it);
+        }
+
+        // This is a safeguard in the case where ancestors of transactions in this block have re-entered
+        // the mempool, possibly from a re-org. This should never happen and would likely indicate a locking
+        // issue if it did.
+        for (txiter it : setAncestorsFromBlock)
+        {
+            DbgAssert("Ancestors in the mempool when they should not be", );
```

Comment:
This DbgAssert is always true.

Reply:
I understand what you are saying here which is that every ancestor tx that's in the mempool MUST be in the block (or the block is invalid since the blockchain will be missing a tx).  SetAncestorsFromBlock is really only used to stop _CalculateMemPoolAncestors from repeatedly adding the same tx into the set.

So its impossible for setAncestorsFromBlock to be not empty here, since we know at this point that the block is valid, and setTxnsInBlock must contain every transaction.

What I think you should do release mode is setAncestorsFromBlock.clear() rather than the above loop and erase.  Looping and calling erase is probably n*log(n) time if the set is implemented as a tree with log(n) access time.  Clear() will be linear or constant time.

If you wanted to keep the check in debug mode, you could preserve the erase and then do a check like DbgAssert(setAncestorsFromBlock.size() == 0, "Ancestor wasn't removed from mempool").  We'd want to panic here in debug mode because this condition would mean that the block wasn't checked properly (all ancestors aren't in the block).  Or, as you commented, the WRITELOCK(cs_txmempool) we did at the top is completely broken.

Reply:
ok, I updated so we only do the dbgassert once if the set is not empty.

Reply:
But its still always true.  I think you want a ! in front so its always false:

DbgAssert(!"Ancestors in the mempool when they should not be", );

Reply:
you're right, ok done.